File: CfxGenerator/CodeBuilder.vb
'' Copyright (c) 2014-2015 Wolfgang Borgsmüller
'' All rights reserved.
'' 
'' Redistribution and use in source and binary forms, with or without 
'' modification, are permitted provided that the following conditions 
'' are met:
'' 
'' 1. Redistributions of source code must retain the above copyright 
''    notice, this list of conditions and the following disclaimer.
'' 
'' 2. Redistributions in binary form must reproduce the above copyright 
''    notice, this list of conditions and the following disclaimer in the 
''    documentation and/or other materials provided with the distribution.
'' 
'' 3. Neither the name of the copyright holder nor the names of its 
''    contributors may be used to endorse or promote products derived 
''    from this software without specific prior written permission.
'' 
'' THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 
'' "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT 
'' LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS 
'' FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE 
'' COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, 
'' INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
'' BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS 
'' OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND 
'' ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR 
'' TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE 
'' USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Public Class CodeBuilder

    Private Const spaces As String = "                                                                                                                        "

    Protected builder As New Text.StringBuilder

    Private currentLine As New List(Of String)

    Public CurrentIndent As Integer

    Sub New()
    End Sub

    Sub New(ByVal indent As Integer)
        CurrentIndent = indent
    End Sub

    Private Sub EmitDebugInfo()
        If Main.EmitDebugInfo > 0 Then
            Dim i = 2
            Dim caller = New StackFrame(2, False)
            Dim method = caller.GetMethod()
            Do While method.DeclaringType.Name = "CodeBuilder" OrElse method.DeclaringType.Name = "ClassBuilder"
                i += 1
                caller = New StackFrame(i, False)
                method = caller.GetMethod()
            Loop
            AppendComment(" {0}.{1}", method.DeclaringType.Name, method.Name)
        End If
    End Sub

    Private Sub CommitCurrentLine()
        If currentLine.Count > 0 Then
            If CurrentIndent > 0 Then builder.Append(spaces.Substring(0, CurrentIndent * 4))
            For Each s In currentLine
                builder.Append(s)
            Next
            currentLine.Clear()
        End If
    End Sub

    Public Sub IncreaseIndent()
        CurrentIndent += 1
    End Sub

    Public Sub DecreaseIndent()
        CurrentIndent -= 1
    End Sub

    Sub BeginBlock(ByVal header As String, ByVal ParamArray pm As Object())
        Append(header, pm)
        Append(" ")
        BeginBlock()
    End Sub

    Sub BeginBlock(ByVal header As String)
        Append(header)
        Append(" ")
        BeginBlock()
    End Sub

    Sub BeginBlock()
        AppendLine("{")
        IncreaseIndent()
    End Sub

    Sub EndBlock()
        DecreaseIndent()
        AppendLine("}")
    End Sub

    Sub EndBlock(ByVal suffix As String)
        DecreaseIndent()
        Append("} ")
        AppendLine(suffix)
    End Sub

    Sub EndBlock(ByVal suffix As String, ByVal ParamArray pm As Object())
        DecreaseIndent()
        Append("} ")
        AppendLine(suffix, pm)
    End Sub

    Sub Append(ByVal line As String)
        currentLine.Add(line)
    End Sub

    Sub AppendLine()
        CommitCurrentLine()
        EmitDebugInfo()
        builder.AppendLine()
    End Sub

    Sub AppendLine(ByVal line As String)
        Append(line)
        AppendLine()
    End Sub

    Sub Append(ByVal format As String, ByVal ParamArray pm As Object())
        Append(String.Format(format, pm))
    End Sub

    Sub AppendLine(ByVal format As String, ByVal ParamArray pm As Object())
        Append(format, pm)
        AppendLine()
    End Sub

    Sub AppendMultiline(ByVal code As String)
        Dim lines = code.Split(Environment.NewLine.ToCharArray())
        For Each line In lines
            AppendLine(line)
        Next
    End Sub

    Sub AppendMultiline(ByVal format As String, ByVal ParamArray pm As Object())
        Dim code = String.Format(format, pm)
        Dim lines = code.Split(Environment.NewLine.ToCharArray())
        For Each line In lines
            AppendLine(line)
        Next
    End Sub

    Sub AppendComment(line As String)
        Append("// ")
        AppendLine(line)
    End Sub

    Sub BeginCfxNamespace()
        BeginCfxNamespace({"System"})
    End Sub

    Sub BeginCfxNamespace(subspace As String)
        BeginCfxNamespace({"System"}, subspace)
    End Sub

    Sub BeginCfxNamespace(usingStmts As String(), Optional subspace As String = "")
        For Each stmt In usingStmts
            AppendLine("using {0};", stmt)
        Next
        AppendLine()
        BeginBlock("namespace Chromium" & subspace)
    End Sub

    Sub BeginClass(name As String, Optional modifiers As String = "public")
        BeginBlock("{0} class {1}", modifiers, name)
    End Sub

    Sub BeginRemoteCallClass(name As String, eventCall As Boolean, callIds As List(Of String))
        If Not eventCall Then
            BeginClass(name & "RenderProcessCall : RenderProcessCall", "internal")
            callIds.Add(name & "RenderProcessCall")
            AppendLine()
            AppendLine("internal {0}RenderProcessCall()", name)
            IncreaseIndent()
            AppendLine(": base(RemoteCallId.{0}RenderProcessCall) {{}}", name)
            DecreaseIndent()
        Else
            BeginClass(name & "BrowserProcessCall : BrowserProcessCall", "internal")
            callIds.Add(name & "BrowserProcessCall")
            AppendLine()
            AppendLine("internal {0}BrowserProcessCall()", name)
            IncreaseIndent()
            AppendLine(": base(RemoteCallId.{0}BrowserProcessCall) {{}}", name)
        DecreaseIndent()
        End If
    End Sub


    Sub BeginFunction(functionSignature As String, Optional modifiers As String = "public")
        BeginBlock("{0} {1}", modifiers, functionSignature)
    End Sub

    Sub BeginFunction(name As String, returnType As String, arguments As String, Optional modifiers As String = "public")
        BeginBlock("{0} {1} {2}({3})", modifiers, returnType, name, arguments)
    End Sub

    Sub BeginIf(condition As String)
        BeginBlock("if({0})", condition)
    End Sub

    Sub BeginIf(condition As String, ByVal ParamArray pm As Object())
        BeginBlock("if({0})", String.Format(condition, pm))
    End Sub

    Sub BeginFor(limit As String)
        BeginBlock("for(int i = 0; i < {0}; ++i)", limit)
    End Sub

    Sub BeginElse()
        EndBlock("else {")
        IncreaseIndent()
    End Sub

    Sub BeginCatch()
        EndBlock("catch {")
        IncreaseIndent()
    End Sub

    Sub BeginCatch(catchArgument As String)
        EndBlock(String.Format("catch({0}) {{", catchArgument))
        IncreaseIndent()
    End Sub

    Sub AppendComment(format As String, ByVal ParamArray pm As Object())
        AppendComment(String.Format(format, pm))
    End Sub

    Sub AppendSummary(summary As CommentData, Optional forRemote As Boolean = False, Optional forEvent As Boolean = False)
        If summary IsNot Nothing AndAlso Not summary.Lines.Length = 0 Then
            AppendLine("/// <summary>")
            For Each line In summary.Lines
                line = CSharp.PrepareSummaryLine(line, forRemote, forEvent)
                AppendLine("/// " & line)
            Next
            AppendLine("/// </summary>")
        End If
    End Sub


    Sub AppendBuilder(ByVal b As CodeBuilder)
        CommitCurrentLine()
        builder.Append(b.ToString())
    End Sub

    Sub CutRight(charCount As Integer)
        CommitCurrentLine()
        builder.Remove(builder.Length - charCount, charCount)
    End Sub

    Sub TrimRight()
        CommitCurrentLine()
        Dim i = builder.Length - 1
        Do While i >= 0 AndAlso Char.IsWhiteSpace(builder.Chars(i))
            i -= 1
        Loop
        If i < builder.Length - 1 Then
            builder.Remove(i + 1, builder.Length - i - 1)
        End If
    End Sub

    ReadOnly Property IsNotEmpty As Boolean
        Get
            Return currentLine.Count > 0 OrElse builder.Length > 0
        End Get
    End Property

    Sub Clear()
        builder.Clear()
    End Sub

    Public Overrides Function ToString() As String
        CommitCurrentLine()
        Return builder.ToString()
    End Function
End Class

File: CfxGenerator/CodeBuilder.cs
// Copyright (c) 2014-2017 Wolfgang Borgsmüller
// All rights reserved.
// 
// This software may be modified and distributed under the terms
// of the BSD license. See the License.txt file for details.

using System.Collections.Generic;
using System.Diagnostics;
using System.Text;
using System.Text.RegularExpressions;

public class CodeBuilder {

    private const string spaces = "                                                                                                                        ";

    protected StringBuilder builder = new StringBuilder();

    private List<string> currentLine = new List<string>();

    public int CurrentIndent;

    public CodeBuilder() {
    }

    public CodeBuilder(int indent) {
        CurrentIndent = indent;
    }

    private void EmitDebugInfo() {
        if(Program.EmitDebugInfo > 0) {
            var i = 2;
            var caller = new StackFrame(2, false);
            var method = caller.GetMethod();
            while(method.DeclaringType.Name == "CodeBuilder" || method.DeclaringType.Name == "ClassBuilder") {
                i += 1;
                caller = new StackFrame(i, false);
                method = caller.GetMethod();
            }
            AppendComment(" {0}.{1}", method.DeclaringType.Name, method.Name);
        }
    }

    private void CommitCurrentLine() {
        if(currentLine.Count > 0) {
            if(CurrentIndent > 0)
                builder.Append(spaces.Substring(0, CurrentIndent * 4));
            foreach(var s in currentLine) {
                builder.Append(s);
            }
            currentLine.Clear();
        }
    }

    public void IncreaseIndent() {
        CurrentIndent += 1;
    }

    public void DecreaseIndent() {
        CurrentIndent -= 1;
    }

    public void BeginBlock(string header, params object[] pm) {
        Append(header, pm);
        Append(" ");
        BeginBlock();
    }

    public void BeginBlock(string header) {
        Append(header);
        Append(" ");
        BeginBlock();
    }

    public void BeginBlock() {
        AppendLine("{");
        IncreaseIndent();
    }

    public void EndBlock() {
        DecreaseIndent();
        AppendLine("}");
    }

    public void EndBlock(string suffix) {
        DecreaseIndent();
        Append("} ");
        AppendLine(suffix);
    }

    public void EndBlock(string suffix, params object[] pm) {
        DecreaseIndent();
        Append("} ");
        AppendLine(suffix, pm);
    }

    public void Append(string line) {
        currentLine.Add(line);
    }

    public void AppendLine() {
        CommitCurrentLine();
        EmitDebugInfo();
        builder.Append("\r\n");
    }

    public void AppendLine(string line) {
        Append(line);
        AppendLine();
    }

    public void Append(string format, params object[] pm) {
        Append(string.Format(format, pm));
    }

    public void AppendLine(string format, params object[] pm) {
        Append(format, pm);
        AppendLine();
    }

    public void AppendMultiline(string code) {
        var lines = Regex.Split(code, "[\r\n]+");
        foreach(var line in lines) {
            AppendLine(line);
        }
    }

    public void AppendMultiline(string format, params object[] pm) {
        var code = string.Format(format, pm);
        var lines = Regex.Split(code, "[\r\n]+");
        foreach(var line in lines) {
            AppendLine(line);
        }
    }

    public void AppendComment(string line) {
        Append("// ");
        AppendLine(line);
    }

    public void BeginCfxNamespace() {
        BeginCfxNamespace(new string[] { "System" });
    }

    public void BeginCfxNamespace(string subspace) {
        BeginCfxNamespace(new string[] { "System" }, subspace);
    }

    public void BeginCfxNamespace(string[] usingStmts, string subspace = "") {
        foreach(var stmt in usingStmts) {
            AppendLine("using {0};", stmt);
        }
        AppendLine();
        BeginBlock("namespace Chromium" + subspace);
    }

    public void BeginClass(string name, string modifiers = "public") {
        BeginBlock("{0} class {1}", modifiers, name);
    }

    public void BeginRemoteCallClass(string name, List<string> callIds, string baseClass = "RemoteCall") {
            BeginBlock("internal class {0}{1} : {1}", name, baseClass);
            callIds.Add(name + baseClass);
            AppendLine();
            AppendLine("internal {0}{1}()", name, baseClass);
            IncreaseIndent();
            AppendLine(": base(RemoteCallId.{0}{1}) {{}}", name, baseClass);
            DecreaseIndent();
    }

    public void BeginFunction(string functionSignature, string modifiers = "public") {
        BeginBlock("{0} {1}", modifiers, functionSignature);
    }

    public void BeginFunction(string name, string returnType, string parameters, string modifiers = "public") {
        BeginBlock("{0} {1} {2}({3})", modifiers, returnType, name, parameters);
    }

    public void BeginIf(string condition) {
        BeginBlock("if({0})", condition);
    }

    public void BeginIf(string condition, params object[] pm) {
        BeginBlock("if({0})", string.Format(condition, pm));
    }

    public void BeginFor(string limit) {
        BeginBlock("for(int i = 0; i < {0}; ++i)", limit);
    }

    public void BeginFor(string limit, params object[] pm) {
        BeginFor(string.Format(limit, pm));
    }

    public void BeginElse() {
        EndBlock("else {");
        IncreaseIndent();
    }

    public void BeginElseIf(string condition, params object[] pm) {
        EndBlock("else if({0}) {{", string.Format(condition, pm));
        IncreaseIndent();
    }

    public void BeginCatch() {
        EndBlock("catch {");
        IncreaseIndent();
    }

    public void BeginCatch(string catchArgument) {
        EndBlock(string.Format("catch({0}) {{", catchArgument));
        IncreaseIndent();
    }

    public void AppendComment(string format, params object[] pm) {
        AppendComment(string.Format(format, pm));
    }

    public void AppendSummary(CommentNode summary, bool forRemote = false, bool forEvent = false) {
        if(summary != null && !(summary.Lines.Length == 0)) {
            AppendLine("/// <summary>");
            foreach(var line in summary.Lines) {
                var l1 = CSharp.PrepareSummaryLine(line, forRemote, forEvent);
                AppendLine("/// " + l1);
            }
            AppendLine("/// </summary>");
        }
    }

    public void AppendSummaryAndRemarks(CommentNode summary, bool forRemote = false, bool forEvent = false) {
        if(summary != null && !(summary.Lines.Length == 0)) {
            AppendSummary(summary, forRemote, forEvent);
            AppendLine("/// <remarks>");
            AppendLine("/// See also the original CEF documentation in");
            AppendLine("/// <see href=\"https://bitbucket.org/chromiumfx/chromiumfx/src/tip/{0}\">{0}</see>.", summary.FileName.Replace("\\", "/"));
            AppendLine("/// </remarks>");
        }
    }

    public void AppendBuilder(CodeBuilder b) {
        CommitCurrentLine();
        builder.Append(b.ToString());
    }

    public void CutRight(int charCount) {
        CommitCurrentLine();
        builder.Remove(builder.Length - charCount, charCount);
    }

    public void TrimRight() {
        CommitCurrentLine();
        var i = builder.Length - 1;
        while(i >= 0 && char.IsWhiteSpace(builder[i])) {
            i -= 1;
        }
        if(i < builder.Length - 1) {
            builder.Remove(i + 1, builder.Length - i - 1);
        }
    }

    public bool IsNotEmpty {
        get { return currentLine.Count > 0 || builder.Length > 0; }
    }

    public void Clear() {
        builder.Clear();
    }

    public override string ToString() {
        CommitCurrentLine();
        return builder.ToString();
    }
}